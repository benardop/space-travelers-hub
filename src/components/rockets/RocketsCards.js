import React from 'react';
import '../stylesheets/RocketsCards.css';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelRocket } from '../../redux/rockets/RocketsSlice';

function RocketsCards({ rocket }) {
  const dispatch = useDispatch();

  const {
    rocketId, rocketName, rocketDesc, rocketImg, rocketReserved,
  } = rocket;

  const reserveHandler = (e) => {
    const data = e.target.id;
    dispatch(reserveRocket(data));
  };

  const cancelHandler = (e) => {
    const data = e.target.id;
    dispatch(cancelRocket(data));
  };

  return (
    <section className="rocketCard">
      <img className="rocketImg" src={rocketImg} alt="" srcSet="" />
      <div className="rocketDesc">
        <h1 className="titleDesc">{rocketName}</h1>
        <p className="paraDesc">
          {rocketReserved === true && (<span className="rocketReserved">Reserved</span>)}
          {rocketDesc}
        </p>
        { rocketReserved === true
          ? (
            <button className="cancelBtn" type="button" id={rocketId} onClick={cancelHandler}>
              Cancel Reservation
            </button>
          )
          : (
            <button className="reserveBtn" type="button" id={rocketId} onClick={reserveHandler}>
              Reserve Rocket
            </button>
          )}
      </div>
    </section>
  );
}

RocketsCards.propTypes = {
  rocket: PropTypes.shape({
    rocketId: PropTypes.string,
    rocketName: PropTypes.string,
    rocketDesc: PropTypes.string,
    rocketImg: PropTypes.string,
    rocketReserved: PropTypes.bool,
  }).isRequired,
};

export default RocketsCards;
