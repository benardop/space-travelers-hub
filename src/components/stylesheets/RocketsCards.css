@import '../../App.css';

.rocketCard {
  display: flex;
  background-color: var(--color-white);
  border: 1px solid grey;
  margin: 0.5rem;
  padding: 0.5rem;
  align-items: flex-start;
}

.rocketImg {
  width: 5rem;
  height: 5rem;
  padding: 0.5rem;
  object-fit: cover;
}

.rocketDesc {
  display: flex;
  flex-direction: column;
  font-size: 0.75rem;
}

.titleDesc {
  font-size: 1rem;
  margin: 0;
}

.paraDesc {
  margin: 0;
}

.rocketReserved {
  display: inline-block;
  background-color: var(--color-greenishblue);
  color: var(--color-white);
  width: fit-content;
  border-radius: 0.5rem;
  padding: 0.5rem;
  font-size: 0.75rem;
  margin-right: 1rem;
}

.reserveBtn {
  background-color: var(--color-blue);
  color: var(--color-white);
  width: 8rem;
  border: none;
  border-radius: 0.5rem;
  padding: 0.5rem 1rem;
  cursor: pointer;
  bottom: 0;
}

.cancelBtn {
  background-color: var(--color-white);
  color: #6d757d;
  width: 10rem;
  border: 1px solid #6d757d;
  border-radius: 0.5rem;
  padding: 0.5rem 0;
  cursor: pointer;
  bottom: 0;
}

.reserveBtn:hover {
  background-color: var(--color-greenishblue);
}

@media screen and (min-width: 768px) {
  .rocketCard {
    width: 90vw;
    margin: 1rem auto;
  }

  .rocketImg {
    height: 200px;
    width: 250px;
    object-fit: fill;
  }

  .rocketDesc {
    font-size: 1rem;
    height: 12rem;
    position: relative;
  }

  .titleDesc {
    font-size: 1.5rem;
  }

  .rocketReserved {
    font-size: 1rem;
  }

  .paraDesc {
    margin: 0;
  }

  .reserveBtn,
  .cancelBtn {
    width: 10rem;
    font-size: 1rem;
    position: absolute;
  }
}
