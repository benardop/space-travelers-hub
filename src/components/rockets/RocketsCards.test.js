import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import store from '../../redux/configureStore';
import RocketsCards from './RocketsCards';

// testing rockets card
const rocket = {
  rocketId: 'falcon1',
  rocketName: 'Falcon 1',
  rocketDesc:
      'The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.',
  rocketImg: 'https://imgur.com/DaCfMsj.jpg',
  rocketWiki: 'https://en.wikipedia.org/wiki/Falcon_1',
  rocketReserved: true,
};

const MockRockets = () => (
  <Provider store={store}>
    <RocketsCards rocket={rocket} />
  </Provider>
);

test('renders correctly', () => {
  const tree = renderer.create(<MockRockets />).toJSON();
  expect(tree).toMatchSnapshot();
});

// test button change text
describe('when you click button it should change to reserved', () => {
  test('should render rockets page content', () => {
    render(
      (<MockRockets />),
    );
    userEvent.click(screen.getByRole('button', { name: /cancel reservation/i }));
    expect(screen.getByText(/reserved/i)).toBeInTheDocument();
  });
});
