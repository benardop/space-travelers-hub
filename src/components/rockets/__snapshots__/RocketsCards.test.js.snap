// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="rocketCard"
>
  <img
    alt=""
    className="rocketImg"
    src="https://imgur.com/DaCfMsj.jpg"
    srcSet=""
  />
  <div
    className="rocketDesc"
  >
    <h1
      className="titleDesc"
    >
      Falcon 1
    </h1>
    <p
      className="paraDesc"
    >
      <span
        className="rocketReserved"
      >
        Reserved
      </span>
      The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
    </p>
    <button
      className="cancelBtn"
      id="falcon1"
      onClick={[Function]}
      type="button"
    >
      Cancel Reservation
    </button>
  </div>
</section>
`;
